app/controllers/vouchers_controller.rb

```diff
@@ -0,0 +1,85 @@
+class VouchersController < ApplicationController
+  before_action :set_voucher, only: [:invalidate]
+  authorize_resource except: [:create, :redeem]
```

Comment:
In addition to the custom `authorize! :create, @voucher`, can't we also use the `authorize_resource` here for `create`? Since we know which people are allowed to even call the endpoint. Therefore, we wouldn't have to unnecessarily create a new voucher and try to save it just to notice: "ups, you are not even allowed to call this method".

Reply:
The problem is that whether a user can create a voucher also depends on the lecture (in a lecture where the user is e.g. an editor, the user can create a voucher, but not in a lecture where the user has no additional permissions).
Therefore it states in the corresponding ability that
```
can [:create, :invalidate], Voucher do |voucher|
    user.can_update_personell?(voucher.lecture)
end
```
In order for that to work, `@voucher` in the controller needs to be set up correctly, in particular with the correct lecture. cancancan hooks right into before the create action (but after the before_actions). But `@voucher`  is only set within the create action. Therefore, with the current code, removing `except :create` and removing the manual authorization in the create action would result in anybody being able to create vouchers. If we would want to remove `except: :create` and the manual authorization for the create action, we would have to define a before action that correctly sets up `@voucher`:

```
before_action :set_new_voucher, only: :create

def set_new_voucher
  @voucher = Voucher.new(voucher_params)
end
```

Reply:
As discussed off-GitHub, there's a [`load_and_authorize_resource`](https://github.com/CanCanCommunity/cancancan/blob/develop/docs/controller_helpers.md#authorize_resource-load_resource-load_and_authorize_resource) method that simplifies this. Added in cb5863dd.